atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/InstrumentedConsensusForgettingStore.java

```diff
@@ -0,0 +1,96 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.pue;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CheckAndSetException;
+import com.palantir.atlasdb.keyvalue.api.KeyAlreadyExistsException;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class InstrumentedConsensusForgettingStore implements ConsensusForgettingStore {
+    private final ConsensusForgettingStore delegate;
+    private final ConsensusForgettingStoreMetrics metrics;
+    private final AtomicInteger concurrentCheckAndTouchOperations;
+
+    @VisibleForTesting
+    InstrumentedConsensusForgettingStore(
+            ConsensusForgettingStore delegate,
+            ConsensusForgettingStoreMetrics metrics,
+            AtomicInteger concurrentCheckAndTouchOperations) {
+        this.delegate = delegate;
+        this.metrics = metrics;
+        this.concurrentCheckAndTouchOperations = concurrentCheckAndTouchOperations;
+    }
+
+    public static ConsensusForgettingStore create(ConsensusForgettingStore delegate, TaggedMetricRegistry registry) {
+        ConsensusForgettingStoreMetrics metrics = ConsensusForgettingStoreMetrics.of(registry);
+        AtomicInteger concurrentCheckAndTouchOperationTracker = new AtomicInteger(0);
+        metrics.concurrentCheckAndTouches(concurrentCheckAndTouchOperationTracker::get);
+        return new InstrumentedConsensusForgettingStore(delegate, metrics, concurrentCheckAndTouchOperationTracker);
+    }
+
+    @Override
+    public void putUnlessExists(Cell cell, byte[] value) throws KeyAlreadyExistsException {
+        delegate.putUnlessExists(cell, value);
+    }
+
+    @Override
+    public void putUnlessExists(Map<Cell, byte[]> values) throws KeyAlreadyExistsException {
+        delegate.putUnlessExists(values);
+    }
+
+    @Override
+    public void checkAndTouch(Cell cell, byte[] value) throws CheckAndSetException {
+        concurrentCheckAndTouchOperations.incrementAndGet();
+        try {
+            metrics.checkAndTouch().time(() -> delegate.checkAndTouch(cell, value));
+        } finally {
+            concurrentCheckAndTouchOperations.decrementAndGet();
+        }
+    }
+
+    @Override
+    public void checkAndTouch(Map<Cell, byte[]> values) throws CheckAndSetException {
+        metrics.batchedCheckAndTouchSize().update(values.size());
```

Comment:
What does the batched metric give us? It is a bit confusing because if you want the number of calls to checkAndTouch, then it seems that should be sum of batchedCalls and direct calls to checkAndTouch. 
I am missing some context here.

Reply:
We'll actually have number of calls to checkAndTouch through the timer. What I want here is to know if someone called the map version with a 200-element map. The problem is that that's a good candidate for batching, but only counting the concurrency of checkAndTouch wouldn't identify that (since the concurrency "limit" would be 1). So this is designed to give an estimate of batch size.